lib/microcbor/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+//! reason to exist.
+//!
+//! [`minicbor-derive`]: https://docs.rs/minicbor-derive
+#![no_std]
+use encode::{Encoder, Write};
+#[doc(inline)]
+pub use microcbor_derive::{Encode, EncodeFields};
+pub use minicbor::encode::{self, Encode};
+
+/// A CBOR-encodable value with a statically-known maximum length.
+///
+/// A type implementing this trait must implement the [`Encode`]`<()>` trait. In
+/// addition, it defines a `MAX_CBOR_LEN` constant that specifies the maximum
+/// number of bytes that its [`Encode`] implementation will produce.
+pub trait StaticCborLen: Encode<()> {
+    /// The maximum length of the CBOR-encoded representation of this value.
+    ///
+    /// The value is free to encode fewer than this many bytes, but may not
+    /// encode more.
+    const MAX_CBOR_LEN: usize;
+}
+
+/// For a list of types implementing [`StaticCborLen`], returns the maximum length
+/// of their CBOR-encoded representations.
+///
+/// This macro may be used to calculate the maximum buffer size necessary to
+/// encode any of a set of types implementing [`StaticCborLen`].
+///
+/// For example:
+///
+/// ```rust
+///
+/// #[derive(microcbor::Encode)]
+/// pub struct MyGreatEreport {
+///     foo: u32,
+///     bar: Option<f64>,
+/// }
+///
+/// #[derive(microcbor::Encode)]
+/// pub enum AnotherEreport {
+///     A { hello: bool, world: f64 },
+///     B(usize),
+/// }
+///
+/// const EREPORT_BUF_LEN: usize = microcbor::max_cbor_len_for![
+///     MyGreatEreport,
+///     AnotherEreport,
+/// ];
+///
+/// fn main() {
+///     let mut ereport_buf = [0; EREPORT_BUF_LEN];
+///     // ...
+///     # drop(ereport_buf);
+/// }
+/// ```
+#[macro_export]
+macro_rules! max_cbor_len_for {
+    ($($T:ty),+$(,)?) => {
+        {
+            let mut len = 0;
+            $(
+                if <$T as $crate::StaticCborLen>::MAX_CBOR_LEN > len {
+                    len = <$T as $crate::StaticCborLen>::MAX_CBOR_LEN;
+                }
+            )+
+            len
+        }
+    };
+}
+
+/// Encode the named fields of a type into an *existing* CBOR map as name-value
+/// pairs.
+///
+/// This is used when a type is included as a field in a "parent" type that
+/// derives `EreportData`, and the field in the parent type is annotated with
+/// `#[ereport(flatten)]`. When that attribute is present, the fields of the
+/// type are encoded as name-value pairs in the parent type's CBOR map, rather
+/// than creating a new nested map for the new type being encoded.
+///
+/// This type may be derived by struct types with named fields, and by enum
+/// types where all variants have named fields.
+pub trait EncodeFields<C> {
+    const MAX_FIELDS_LEN: usize;
+
+    fn encode_fields<W: Write>(
+        &self,
+        e: &mut Encoder<W>,
+        _: &mut C,
+    ) -> Result<(), encode::Error<W::Error>>;
+}
+
+#[cfg(feature = "fixedstr")]
+impl<const LEN: usize> StaticCborLen for fixedstr::FixedStr<LEN> {
```

Comment:
Why put this in `microcbor`, instead of having an optional `microcbor` dependency in `fixedstr` (like it already has for `minicbor` and `serde`)?